src/hotspot/share/nmt/nmtCommon.hpp

```diff
@@ -137,5 +141,32 @@ class NMTUtil : AllStatic {
   static S _strings[mt_number_of_types];
 };
 
+class NmtGuard : public StackObj {
+private:
+    static Semaphore _nmt_semaphore;
+    static intx volatile _owner;
+
+public:
+    NmtGuard() {
+      intx const current =  os::current_thread_id();
+      intx const owner = Atomic::load(&_owner);
+      assert(current != owner, "Lock is not reentrant");
+
+      _nmt_semaphore.wait();
+      Atomic::store(&_owner, current);
+    }
+
+    ~NmtGuard() {
+      assert_locked();
+      Atomic::store(&_owner, (intx) -1);
+      _nmt_semaphore.signal();
+    }
+
+    static void assert_locked(){
+      intx const current = os::current_thread_id();
+      intx const owner = Atomic::load(&_owner);
+      assert(current == owner, "NMT lock should be acquired in this section.");
```

Comment:
Please put the entire body in `ifdef DEBUG` so we don't execute anything in release builds.

Also the assert should report the values of owner and current when they are not as expected.